testsuite/tests/typing-gadts/pr10189.ml

```diff
@@ -0,0 +1,191 @@
+(* TEST
+   * expect
+*)
+
+type i = <m : 'c. 'c -> 'c >
+type ('a, 'b) j = <m : 'c. 'a -> 'b >
+type _ t = A : i t;;
+[%%expect{|
+type i = < m : 'c. 'c -> 'c >
+type ('a, 'b) j = < m : 'a -> 'b >
+type _ t = A : i t
+|}]
+
+let f (type a b) (y : (a, b) j t) : a -> b =
+  let A = y in fun x -> x;;
+[%%expect{|
+Line 2, characters 6-7:
+2 |   let A = y in fun x -> x;;
+          ^
+Error: This pattern matches values of type i t
+       but a pattern was expected which matches values of type (a, b) j t
+       Type i = < m : 'c. 'c -> 'c > is not compatible with type
+         (a, b) j = < m : a -> b >
+       The method m has type 'c. 'c -> 'c, but the expected method type was
+       a -> b
+       The universal variable 'c would escape its scope
+|}]
+
+let g (type a b) (y : (a,b) j t option) =
+  let None = y in () ;;
+[%%expect{|
+val g : ('a, 'b) j t option -> unit = <fun>
+|}]
+
+module M = struct
+  type 'a d = D
+  type j = <m : 'c. 'c -> 'c d >
+end ;;
+let g (y : M.j t option) =
+  let None = y in () ;;
+[%%expect{|
+module M : sig type 'a d = D type j = < m : 'c. 'c -> 'c d > end
+val g : M.j t option -> unit = <fun>
+|}]
+
+module M = struct
+  type 'a d
+  type j = <m : 'c. 'c -> 'c d >
+end ;;
+let g (y : M.j t option) =
+  let None = y in () ;;
+[%%expect{|
+module M : sig type 'a d type j = < m : 'c. 'c -> 'c d > end
+Line 6, characters 2-20:
+6 |   let None = y in () ;;
+      ^^^^^^^^^^^^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+Some A
+val g : M.j t option -> unit = <fun>
+|}]
+
+module M = struct
+  type e
+  type 'a d
+  type i = <m : 'c. 'c -> 'c d >
+  type j = <m : 'c. 'c -> e >
+end ;;
+type _ t = A : M.i t
+let g (y : M.j t option) =
+  let None = y in () ;;
+[%%expect{|
+module M :
+  sig
+    type e
+    type 'a d
+    type i = < m : 'c. 'c -> 'c d >
+    type j = < m : 'c. 'c -> e >
+  end
+type _ t = A : M.i t
+Line 9, characters 2-20:
+9 |   let None = y in () ;;
+      ^^^^^^^^^^^^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
```

Comment:
Shouldn't the `Some A` case here be ruled out? I think you'll need to add an `occur_univar` check to instantiable `Tconstr`s in `unify3`.

Reply:
Just add `type 'a d = e` in the above piece of code, and you can see that the `Some A` may really happen. We have to be extremely careful with those (and conservative when we don't know for sure).